### DISCUSSION TITLE
Glfw and Opengl. ImGui rendering code is executed without errors, but no UI elements are shown

### DISCUSSION THREAD

**Discussion by uzername** [ocornut/imgui#5968] [New Users Build/Link/Run issues ONLY!]
Hello. I am trying to integrate Dear ImGui code with OpenCascade - an engineering library for showing 3D elements which runs on OpenGL and inside window provided by GLFW. I took an example from their archive, which is dedicted to GLFW rendering, I provide reference to GLFW window but no interface elements are shown. Where should I look in debugger? What may be causing such issue? 
My main subroutine looks like this:
```
void GlfwOcctView::run()
{
  initWindow (800, 600, "glfw occt");
  initViewer();
  initDemoScene();
  if (myView.IsNull())
  {
    return;
  }

  myView->MustBeResized();
  myOcctWindow->Map();
  initUI();
  mainloop();
  cleanupUI();
  cleanup();
}
```
These are subroutines which call glfw init and set up opengl context:
```
// ================================================================
// Function : initWindow
// Purpose  : setup glfw window
// ================================================================
void GlfwOcctView::initWindow (int theWidth, int theHeight, const char* theTitle)
{
  glfwSetErrorCallback (GlfwOcctView::errorCallback);
  glfwInit();
  const bool toAskCoreProfile = true;
  if (toAskCoreProfile)
  {
    glfwWindowHint (GLFW_CONTEXT_VERSION_MAJOR, 3);
    glfwWindowHint (GLFW_CONTEXT_VERSION_MINOR, 3);
#if defined (__APPLE__)
    glfwWindowHint (GLFW_OPENGL_FORWARD_COMPAT, GL_TRUE);
#endif
    glfwWindowHint (GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);
  }
  myOcctWindow = new GlfwOcctWindow (theWidth, theHeight, theTitle);
  glfwSetWindowUserPointer       (myOcctWindow->getGlfwWindow(), this);
  // window callback
  glfwSetWindowSizeCallback      (myOcctWindow->getGlfwWindow(), GlfwOcctView::onResizeCallback);
  glfwSetFramebufferSizeCallback (myOcctWindow->getGlfwWindow(), GlfwOcctView::onFBResizeCallback);
  // mouse callback
  glfwSetScrollCallback          (myOcctWindow->getGlfwWindow(), GlfwOcctView::onMouseScrollCallback);
  glfwSetMouseButtonCallback     (myOcctWindow->getGlfwWindow(), GlfwOcctView::onMouseButtonCallback);
  glfwSetCursorPosCallback       (myOcctWindow->getGlfwWindow(), GlfwOcctView::onMouseMoveCallback);
}

// ================================================================
// Function : initViewer
// Purpose  : make OpenCascade redirect its render to opengl context
// ================================================================
void GlfwOcctView::initViewer()
{
  if (myOcctWindow.IsNull()
   || myOcctWindow->getGlfwWindow() == nullptr)
  {
    return;
  }

  Handle(OpenGl_GraphicDriver) aGraphicDriver = new OpenGl_GraphicDriver (myOcctWindow->GetDisplay(), false);
  Handle(V3d_Viewer) aViewer = new V3d_Viewer (aGraphicDriver);
  aViewer->SetDefaultLights();
  aViewer->SetLightOn();
  aViewer->SetDefaultTypeOfView (V3d_PERSPECTIVE);
  aViewer->ActivateGrid (Aspect_GT_Rectangular, Aspect_GDM_Lines);
  myView = aViewer->CreateView();
  myView->SetImmediateUpdate (false);
  myView->SetWindow (myOcctWindow, myOcctWindow->NativeGlContext());
  myView->ChangeRenderingParams().ToShowStats = false;
  myContext = new AIS_InteractiveContext (aViewer);
}
```
initDemoScene(); contains Opencascade-specific code, not showing it here. By this time the OpenGL and GLFW has been set up

In initUI I set up Dear ImGui, initial config.
```
void GlfwOcctView::initUI()
{
    // Setup Dear ImGui context
    const char* glsl_version = "#version 330";
    IMGUI_CHECKVERSION();
    ImGui::CreateContext();
    ImGuiIO& io = ImGui::GetIO();
    // Setup Platform/Renderer bindings
    ImGui_ImplGlfw_InitForOpenGL(myOcctWindow->getGlfwWindow(), true);
    ImGui_ImplOpenGL3_Init(glsl_version);
    // Setup Dear ImGui style
    ImGui::StyleColorsLight();
}
```
mainloop() handles input and iterates while glfw window has not been closed
```
// ================================================================
// Function : mainloop
// Purpose  :
// ================================================================
void GlfwOcctView::mainloop()
{
  while (!glfwWindowShouldClose (myOcctWindow->getGlfwWindow()))
  {
    // glfwPollEvents() for continuous rendering (immediate return if there are no new events)
    // and glfwWaitEvents() for rendering on demand (something actually happened in the viewer)
    //glfwPollEvents();
    glfwWaitEvents();
    if (!myView.IsNull())
    {
        processUI();
      FlushViewEvents (myContext, myView, true);
    }
    
  }
}
```
FlushViewEvents() is a function of OpenCascade engine, it performs swap buffers, judging from its description:
```
//! Update buffer for rendering thread.
  //! This method is expected to be called within synchronization barrier between GUI
  //! and Rendering threads (e.g. GUI thread should be locked beforehand to avoid data races).
  //! @param theCtx interactive context
  //! @param theView active view
  //! @param theToHandle if TRUE, the HandleViewEvents() will be called
  Standard_EXPORT virtual void FlushViewEvents (const Handle(AIS_InteractiveContext)& theCtx,
                                                const Handle(V3d_View)& theView,
                                                Standard_Boolean theToHandle = Standard_False);
```
In processUI() I call Dear ImGUI routines:
```
void GlfwOcctView::processUI()
{
    // feed inputs to dear imgui, start new frame 
    ImGui_ImplOpenGL3_NewFrame();
    ImGui_ImplGlfw_NewFrame();
    ImGui::NewFrame();
    
    ImGui::Begin("STEP files");   
    
    ImGui::Button("Add File");
    ImGui::SameLine();
    ImGui::Button("Clear List");
    //ImGui::ListBox("Files",&currentItem,listboxItems,currentItemsCount,10);
    
    ImGui::End();

    // Rendering
    ImGui::Render();
    ImGui_ImplOpenGL3_RenderDrawData(ImGui::GetDrawData());
}
```
It is being called, but nothing is rendered as result.
Version/Branch of Dear ImGui:

Version: 1.89.2 VIP

+ Branch: master. I used download ZIP

Back-end/Renderer/Compiler/OS

+ Back-ends: imgui_impl_opengl3.cpp + imgui_impl_glfw.cpp

+ Compiler: Visual Studio 2022

+ Operating System: Win10

**Comment by uzername** [comment]
Apparently, it's more related to OpenCascade rendering cycle. Imgui and OpenCascade did not go well together in my setup, so I had to adjust source code as suggested here: https://dev.opencascade.org/content/using-opencascade-dear-imgui-and-glfw . It has an in-depth and detailed overview of using Opencascade and Dear ImGui. If that thread gets lost in "tides of Internet", here are the changes which gave a result:

> In initViewer() I added this call:
> 
> aGraphicDriver->SetBuffersNoSwap(true);
> And then I changed mainloop() as advised:
> 
> void GlfwOcctView::mainloop()
> {
>   while (!glfwWindowShouldClose (myOcctWindow->getGlfwWindow()))
>   {
>     // glfwPollEvents() for continuous rendering (immediate return if there are no new events)
>     // and glfwWaitEvents() for rendering on demand (something actually happened in the viewer)
>     //glfwPollEvents();
>     glfwWaitEvents();
>     if (!myView.IsNull())
>     {                      
>       FlushViewEvents(myContext, myView, true);
>       processUI();
>       glfwSwapBuffers(myOcctWindow->getGlfwWindow());
>     }
>     
>   }
> }
> There is a little problem though: when moving ImGui window, my scene reacts too on moves. It's worth investigating too

**Comment by ocornut** [comment]
> There is a little problem though: when moving ImGui window, my scene reacts too on moves. It's worth investigating too 

You should ignore inputs events when io.WantCaptureMouse is true. See FAQ:
https://github.com/ocornut/imgui/blob/master/docs/FAQ.md#q-how-can-i-tell-whether-to-dispatch-mousekeyboard-to-dear-imgui-or-my-application
